db/bbolt.go

```diff
@@ -174,6 +166,39 @@ func (bb *BBoltHandler) GetBlockByHash(hash string) (*types.Block, error) {
 	return bb.GetBlockByHeight(blockHeight)
 }
 
+func (bb *BBoltHandler) QueryIsBlockRangeFinalizedByHeight(startHeight, endHeight uint64) ([]bool, error) {
```

Comment:
Would it be possible that block `h` is finalised but block `h+1` is not? I think it's not possiible as BTC staking enforces consecutive finalisation. So wdyt of maintaining the last finalised height in DB and accessing that value, instead of iterating over all these blocks?

Reply:
> Would it be possible that block h is finalised but block h+1 is not?

no it's not possible

> So wdyt of maintaining the last finalised height in DB and accessing that value, instead of iterating over all these blocks?

good point. we already maintain fist and last finalized heights. we can use them

```
const (
	blocksBucket          = "blocks"
	blockHeightsBucket    = "block_heights"
	indexerBucket         = "indexer"
	earliestBlockKey      = "earliest"
	latestBlockKey        = "latest"
	activatedTimestampKey = "activated_timestamp"
)
```

Reply:
Yes, if using last finalised height then the logic can be greatly simplified. Can be done in a separate PR though